File: common/src/main/java/com/goexp/common/db/mongo/DBQueryTemplate.java
package com.goexp.common.db.mongo;

import com.mongodb.client.FindIterable;
import com.mongodb.client.MongoCollection;
import org.bson.Document;
import org.bson.conversions.Bson;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

public class DBQueryTemplate<T> extends AbstractDBTemplate {

    private final MongoCollection<Document> collection;
    private final ObjectCreator<T> defaultCreator;
    private Bson defaultSort;
    private Bson defaultSelect;

    private DBQueryTemplate(String dbName, String tableName, ObjectCreator<T> defaultCreator) {
        super(dbName, tableName);

        Objects.requireNonNull(defaultCreator);
        this.defaultCreator = defaultCreator;

        var db = mongoClient.getDatabase(dbName);
        collection = db.getCollection(tableName);
    }

    private DBQueryTemplate(String dbName, String tableName, ObjectCreator<T> defaultCreator, Bson defaultSelect, Bson defaultSort) {
        this(dbName, tableName, defaultCreator);

        this.defaultSelect = defaultSelect;
        this.defaultSort = defaultSort;
    }

    public QueryBuilder query() {
        return this.new QueryBuilder();
    }

    public static class Builder<T> {
        private String dbName;
        private String tableName;
        private ObjectCreator<T> creator;
        private Bson defaultSort;
        private Bson defaultSelect;


        public Builder(String dbName, String tableName, ObjectCreator<T> creator) {
            this.dbName = dbName;
            this.tableName = tableName;
            this.creator = creator;
        }

        public Builder<T> defaultSort(Bson defaultSort) {
            this.defaultSort = defaultSort;
            return this;
        }

        public Builder<T> defaultSelect(Bson defaultSelect) {
            this.defaultSelect = defaultSelect;
            return this;
        }

        public DBQueryTemplate<T> build() {
            return new DBQueryTemplate<>(dbName, tableName, creator, defaultSelect, defaultSort);
        }
    }

    public class QueryBuilder {

        private Bson where;
        private Bson select;
        private Bson sort;

        private QueryBuilder() {
        }

        public QueryBuilder where(Bson where) {
            this.where = where;
            return this;
        }

        public QueryBuilder select(Bson select) {
            this.select = select;
            return this;
        }

        public QueryBuilder sort(Bson sort) {
            this.sort = sort;
            return this;
        }

        public List<T> list() {
            return getList(defaultCreator);
        }

        public List<T> list(ObjectCreator<T> userCreator) {
            Objects.requireNonNull(userCreator);

            return getList(userCreator);
        }

        public T one() {

            var temp = collection.find();

            // choice where
            if (where != null) {
                temp = temp.filter(where);
            }

            return temp.limit(1).map(defaultCreator::create).first();
        }

        public T one(ObjectCreator<T> userCreator) {
            Objects.requireNonNull(userCreator);

            var temp = collection.find();

            // choice where
            if (where != null) {
                temp = temp.filter(where);
            }

            return temp.limit(1).map(userCreator::create).first();
        }

        public boolean exists() {
            return collection.countDocuments(where) > 0;
        }

        private ArrayList<T> getList(ObjectCreator<T> userCreator) {
            return getDocuments().map(userCreator::create).into(new ArrayList<>());
        }

        private FindIterable<Document> getDocuments() {


            var temp = collection.find();

            // choice where
            if (where != null) {
                temp = temp.filter(where);
            }

            // choice select
            if (select != null)
                temp = temp.projection(select);
            else if (defaultSelect != null)
                temp = temp.projection(defaultSelect);

            // choice sort
            if (sort != null)
                temp = temp.sort(sort);
            else if (defaultSort != null)
                temp = temp.sort(defaultSort);

            return temp;
        }

    }
}

File: common_db_mongo/src/main/scala/com/goexp/common/db/mongo/DBQueryTemplate.scala
package com.goexp.common.db.mongo

import java.util
import java.util._

import org.bson.conversions.Bson

object DBQueryTemplate {

  class Builder[T](private[this] val dbName: String,
                   private[this] val tableName: String,
                   private[this] val creator: ObjectCreator[T]
                  ) {
    private[this] var defaultSort: Bson = _
    private[this] var defaultSelect: Bson = _

    def defaultSort(defaultSort: Bson): Builder[T] = {
      this.defaultSort = defaultSort
      this
    }

    def defaultSelect(defaultSelect: Bson): Builder[T] = {
      this.defaultSelect = defaultSelect
      this
    }

    def build = new DBQueryTemplate[T](dbName, tableName, creator, defaultSelect, defaultSort)
  }

}

class DBQueryTemplate[T] private(dbName: String,
                                 tableName: String,
                                 private[this] val defaultCreator: ObjectCreator[T]
                                ) extends AbstractDBTemplate(dbName, tableName) {
  Objects.requireNonNull(defaultCreator)

  private val collection = mongoClient.getDatabase(dbName).getCollection(tableName)
  private var defaultSort: Bson = _
  private var defaultSelect: Bson = _

  def this(dbName: String,
           tableName: String,
           defaultCreator: ObjectCreator[T],
           defaultSelect: Bson,
           defaultSort: Bson) {
    this(dbName, tableName, defaultCreator)
    this.defaultSelect = defaultSelect
    this.defaultSort = defaultSort
  }

  def query = new QueryBuilder


  class QueryBuilder private[DBQueryTemplate] {
    private[this] val part = new PartBuilder
    private[this] val finalPart = new FinalBuilder

    def where(where: Bson): QueryBuilder = {
      part.where(where)
      this
    }

    def select(select: Bson): QueryBuilder = {
      part.select(select)
      this
    }

    def sort(sort: Bson): QueryBuilder = {
      part.sort(sort)
      this
    }

    def set: util.Set[T] = finalPart.set

    def set(userCreator: ObjectCreator[T]) = {
      Objects.requireNonNull(userCreator)
      finalPart.set(userCreator)
    }

    def list: util.List[T] = finalPart.list

    def list(userCreator: ObjectCreator[T]) = {
      Objects.requireNonNull(userCreator)
      finalPart.list(userCreator)
    }

    def one: T = finalPart.one

    def one(userCreator: ObjectCreator[T]): T = {
      Objects.requireNonNull(userCreator)
      finalPart.one(userCreator)
    }

    def exists: Boolean = finalPart.exists


    class PartBuilder private[QueryBuilder] {
      private var where: Bson = _
      private var select: Bson = _
      private var sort: Bson = _

      def where(where: Bson): PartBuilder = {
        this.where = where
        this
      }

      def select(select: Bson): PartBuilder = {
        this.select = select
        this
      }

      def sort(sort: Bson): PartBuilder = {
        this.sort = sort
        this
      }

      def buildFileIterrableMany = {
        var temp = collection.find()
        if (where != null) temp = temp.filter(where)
        // choice select
        if (select != null) temp = temp.projection(select)
        else if (defaultSelect != null) temp = temp.projection(defaultSelect)
        // choice sort
        if (sort != null) temp = temp.sort(sort)
        else if (defaultSort != null) temp = temp.sort(defaultSort)
        temp
      }

      def buildFileIterrableOne = {
        var temp = collection.find
        if (where != null) temp = temp.filter(where)
        temp.limit(1)
      }

      def getDocumentCount = {
        collection.countDocuments(where)
      }
    }

    class FinalBuilder private[QueryBuilder] {
      def set: util.Set[T] = docs2Collection(defaultCreator.create, new util.HashSet[T])

      def set(userCreator: ObjectCreator[T]) = {
        Objects.requireNonNull(userCreator)
        docs2Collection(userCreator.create, new util.HashSet[T])
      }

      def list: util.List[T] = docs2Collection(defaultCreator.create, new util.ArrayList[T])

      def list(userCreator: ObjectCreator[T]) = {
        Objects.requireNonNull(userCreator)
        docs2Collection(userCreator.create, new util.ArrayList[T])
      }

      def one: T = {
        part.buildFileIterrableOne.map(defaultCreator.create).first
      }

      def one(userCreator: ObjectCreator[T]): T = {
        Objects.requireNonNull(userCreator)
        part.buildFileIterrableOne.map(userCreator.create).first
      }

      def exists: Boolean = part.getDocumentCount > 0


      private def docs2Collection[A <: util.Collection[T]](userCreator: ObjectCreator[T], clazz: A) =
        part.buildFileIterrableMany.map(userCreator.create).into(clazz)

    }

  }

}